### ISSUE TITLE
File Uploads topic enhancements

### ISSUE THREAD

**Issue by guardrex** [aspnet/AspNetCore.Docs#12285]
It won't be possible to address one of the issues in the updates for https://github.com/aspnet/AspNetCore.Docs/issues/10214. 

The following opportunities remain:

* There's no example for handling file Chunking.
* There's no example for Byte Range when downloading.
* Provide help/examples/tips for resuming large file uploads when users were in remote areas with flaky Internet. Ditching streaming large files and going with chunking instead. Differences; pros and cons; terminology explained; common pitfalls.
* RE: Chunked transfer encoding: HTTP/2 apparently doesn't support it ... it goes to DATA frames. Should DATA frames coverage be explained/demoed?

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 9a5cfa38-0e6c-4346-dafc-84ea5c513fcf
* Version Independent ID: c6787f0f-3e9a-6efb-0e6e-5ccb957d3f4a
* Content: [File uploads in ASP.NET Core](https://docs.microsoft.com/en-us/aspnet/core/mvc/models/file-uploads?view=aspnetcore-2.2#feedback)
* Content Source: [aspnetcore/mvc/models/file-uploads.md](https://github.com/aspnet/AspNetCore.Docs/blob/master/aspnetcore/mvc/models/file-uploads.md)
* Product: **aspnet-core**
* Technology: **aspnetcore-mvc**
* GitHub Login: @ardalis
* Microsoft Alias: **riande**

**Comment by nemtajo** [comment]
I am trying to make File Upload For Large Files work in my ASP.NET Core 3.0 Preview project. I followed the tutorial and I am running my project on localhost at the moment. After clicking on Upload button on the razor page, I get:

The SPA default page middleware could not return the default page '/index.html' because it was not found, and no other middleware handled the request.

Maybe the default VS template doesn't have all the necessary configuration in Startup.cs.
I saw that many developers reported this error before:
https://github.com/aspnet/JavaScriptServices/issues/1514
https://github.com/aspnet/AspNetCore/issues/5223
https://github.com/dotnet/core/issues/1803

It would be great if we the examples from github were updated, to have views with .razor extension and ASP.NET Core 3.0 structure:
WebApp.Client
WebApp.Server
WebApp.Shared

I saw the example https://github.com/aspnet/AspNetCore.Docs/blob/ad16f5e1da6c04fa4996ee67b513f2a90fa0d712/aspnetcore/mvc/models/file-uploads/sample/FileUploadSample/Views/Home/Index.cshtml
but I still don't know what is causing the above mentioned default page error.

**Comment by guardrex** [comment]
Hello @nemtajo ... We don't have the staffing to work with projects. We only maintain documentation here. However, there are a **_large number of updates_** on a PR that's in review now. You could take a look at the updates on the PR to see if they would help you. https://github.com/aspnet/AspNetCore.Docs/pull/12344

Otherwise, we recommend talking to devs on a support forum, such as Stack Overflow, or a support chat, such as Slack or Gitter. If you find a real bug in the framework, then open an issue with engineering to report it.

**Comment by nickalbrecht** [comment]
Just noticed the "Master issue" that originally referenced the issue this came from got closed, glad that these are still being looked at/considered though. I haven't forgotten about it, and look forward to documentation enhancements covering uploads!

**Comment by guardrex** [comment]
@nickalbrecht The last round went in today on https://github.com/aspnet/AspNetCore.Docs/pull/14770, and those updates are live now, which include a 3.0 sample.

**Comment by guardrex** [comment]
Additional possible updates are being collected on ... https://github.com/aspnet/AspNetCore.Docs/issues/12285

**Comment by nickalbrecht** [comment]
Good to know. I do think it's important to get some sort of documentation for theses items, some more than others. So I just want to make sure they are not dropped as things not having a high enough ROI for time spent getting them going, or something like that.

Actually, most of theses were carried over multiple times from comment you'd put on https://github.com/aspnet/AspNetCore.Docs/issues/4777#issuecomment-357405456 after a conversation we had on Slack. Few more months, and by Jan 2020 we'll hit 2 years for this one ;-)

I saw an tweet by Fowler the other day about playing with HTTP/3, no idea what potential topics that would bring beyond the original question of chunking and HTTP/2. 😅 But it's obviously still early days for HTTP/3 anyway.

**Comment by guardrex** [comment]
@pranavkm Would you help me prioritize the following (High, Medium, Low, or Hold)?

1. There's no example for handling file chunking; for example to resume large file uploads when users are in remote areas with flaky Internet.
2. HTTP/2 apparently doesn't support chunked transfer encoding ... it goes to DATA frames. Should DATA frames coverage be explained/demoed?
3. There's no example for Byte Range when downloading.
4. The article doesn't mention the `[FileExtensions]` attribute, which can be used to restrict what files can be uploaded. #13905

**Comment by pranavkm** [comment]
1) Sounds really useful. I'd put this at the top
2). @Tratcher, is this useful information for users to understand?
3) I'd put this at number 3
4) Pri 2.This would make sense to cover as part of the file uploads topic.

**Comment by Tratcher** [comment]
> 2\. HTTP/2 apparently doesn't support chunked transfer encoding ... it goes to DATA frames. Should DATA frames coverage be explained/demoed?

Is there an existing chunked transfer encoding sample/explanation you're comparing this to? Chunked encoding and HTTP/2 data frames are wire level framing details, they end up having the same behavior. Any existing description could be generalized to cover both.

**Comment by guardrex** [comment]
@Tratcher No, I'm merely asking about these at this point. Just need to know which items are candidates for enhancing the topic and what priority to apply to them. I understand your remark, so dropping that entry. Should a chunking approach/sample *not* ever be part of this topic (i.e., let blogs/forum post resources cover it)?

**Comment by Tratcher** [comment]
File chunking across multiple requests is a useful technique. Chunked encoding on a single request/response is a formatting detail unrelated the prior scenario. A sentence or two about using or excluding the Content-length header would be sufficient. E.g. Using a content-length header can make the message easier to inspect. When not using a content-length header the message format automatically falls back to transfer-encoding: chunked (or HTTP/2 data frames).

**Comment by nickalbrecht** [comment]
So based off of what you've mentioned regarding the difference between chunked uploads (just split into multiple upload operations), and chunked encoding, Which one is used in the `StreamController` for demonstrating streamed uploads in 
https://github.com/aspnet/AspNetCore.Docs/blob/master/aspnetcore/mvc/models/file-uploads/samples/3.x/SampleApp/Controllers/StreamingController.cs

Is the behaviour of streaming an upload like that the same regardless of the protocol being HTTP/2 or 1.1? I only ask because at first glance it seems to be closer to the underlying details of the protocol than most file upload examples I've seen, and I want to make sure if HTTP/2 implementation details are truly irrelevant regardless of how someone tries to implement file uploading, or if at worst it becomes an edge case that is beyond the intended scope of the documentation and gets mentioned as a foot note saying as such, but that the reader may need to be aware of it if they are looking intend on going down that path.

**Comment by Tratcher** [comment]
> Which one is used in the StreamController for demonstrating streamed uploads in

That shows a file upload on a single request. The request in that example could use Content-Length or Transfer-Encoding: chunked and the sample would behave the same.

> Is the behaviour of streaming an upload like that the same regardless of the protocol being HTTP/2 or 1.1?

Yes. The wire formatting and some headers vary but the application visible behavior does not. E.g. the `Transfer-Encoding: chunked` header is not used in HTTP/2, but the results are the same.

**Comment by gbjbaanb** [comment]
FYI the default nginx upload limit is 1 MB.

To change this for every site in a nginx configuration, edit the nginx.conf and add a 'client_max_body_size' entry to the http section. To edit it per-site, put the entry in the appropriate location section.

http://nginx.org/en/docs/http/ngx_http_core_module.html#client_max_body_size

**Comment by M-Yankov** [comment]
I will try answer the question to (@YehudahA):
> Q: What is the benefit of streaming? #18070

As the article says:
> Any single buffered file exceeding 64 KB is moved from memory to a temp file on disk.

I think one of the benefits is that on high peaks of the applications the file storage will not run out of memory. To observe that the examples from _(AspNetCore.Docs\aspnetcore\mvc\models\file-uploads\samples\3.x\SampleApp)_ can be used. The page `BufferedSingleFileUploadDb` works with single file and a breakpoint in _OnPostUploadAsync() method_ will show the temporary location path.
Using `Shift + F9` I've checked the `FileUpload.FormFile -> ._baseStream -> TempFileName`, it shows: "C:\\Users\\<username>\\AppData\\Local\\Temp\\ASPNETCORE_f08899f8-b3eb-4410-924d-1f44a7b6cc37.tmp" . The file is really there (maybe it's compressed), but that mean if the logic hangs to process requests with large files and there is a lot of usage in the application, a problem with the storage may appear.

On the other side, the streaming approach _(StreamedSingleFileUploadDb page)_ is not accessing the Request.Form.Files or (IFormFile) and thus it's not saved in the temp directory. If I put `var files = this.Request.Form.Files;` in the beginning of the method body, a temp file is created when the execution reaches the new line. In any other case the streaming approach doesn't make temp files.

**Comment by CraigComeOnThisNameCantBeTaken** [comment]
In the sample apps the streamed files are processed by reading the stream in order to determine the file size. Is there still a benefit please?
